ledger/metrics/src/main/scala/com/daml/metrics/grpc/GrpcServerInterceptor.scala

```diff
@@ -0,0 +1,130 @@
+
+class GrpcServerInterceptor(metrics: GrpcServerMetrics) extends ServerInterceptor {
+
+  override def interceptCall[ReqT, RespT](
+      serverCall: ServerCall[ReqT, RespT],
+      metadata: Metadata,
+      serverCallHandler: ServerCallHandler[ReqT, RespT],
+  ): ServerCall.Listener[ReqT] = {
+    val methodDescriptor = serverCall.getMethodDescriptor
+    val clientType = if (methodDescriptor.getType.clientSendsOneMessage()) "unary" else "streaming"
+    val serverType = if (methodDescriptor.getType.serverSendsOneMessage()) "unary" else "streaming"
+    withMetricLabels(
+      "grpc_service_name" -> methodDescriptor.getServiceName,
+      "grpc_method_name" -> methodDescriptor.getBareMethodName,
+      "grpc_client_type" -> clientType,
+      "grpc_server_type" -> serverType,
+    ) { implicit mc =>
+      val timerHandle = metrics.callTimer.startAsync()
+      val metricsServerCall = new MetricsCall(
+        serverCall,
+        timerHandle,
+        metrics.messagesSent,
+        metrics.messagesSentSize,
+        metrics.callsFinished,
+        metrics.messagesRequested,
+      )
+      new MonitoringServerCallListener(
+        serverCallHandler.startCall(metricsServerCall, metadata),
+        metrics.messagesReceived,
+        metrics.messagesReceivedSize,
+        metrics.callsStarted,
+      )
+    }
+  }
+
+  private final class MonitoringServerCallListener[T](
+      val delegate: Listener[T],
+      messagesReceived: Meter,
+      messagesReceivedSize: Histogram,
+      callsStarted: Meter,
+  )(implicit
+      metricsContext: MetricsContext
+  ) extends ForwardingServerCallListener[T] {
+
+    override def onMessage(message: T): Unit = {
+      super.onMessage(message)
+      updateHistogramWithSerializedSize(messagesReceivedSize, message)
+      messagesReceived.mark()
+    }
+    override def onReady(): Unit = {
+      super.onReady()
+      callsStarted.mark()
+    }
+  }
+
+  private final class MetricsCall[ReqT, RespT](
+      call: ServerCall[ReqT, RespT],
+      timer: TimerHandle,
+      messagesSent: Meter,
+      messagesSentSize: Histogram,
+      callsClosed: Meter,
+      messagesRequested: Meter,
+  )(implicit metricsContext: MetricsContext)
+      extends SimpleForwardingServerCall[ReqT, RespT](call) {
+
+    override def request(numMessages: Int): Unit = {
+      super.request(numMessages)
+      messagesRequested.mark(numMessages.toLong)
+    }
+
+    override def sendMessage(message: RespT): Unit = {
+      super.sendMessage(message)
+      updateHistogramWithSerializedSize(messagesSentSize, message)
+      messagesSent.mark()
+    }
+
+    override def close(status: Status, trailers: Metadata): Unit = {
+      super.close(status, trailers)
+      withExtraMetricLabels("code" -> status.getCode.value().toString) { implicit metricsContext =>
+        callsClosed.mark()
+        timer.stop()
+      }
+    }
+  }
+
+  private def updateHistogramWithSerializedSize[T](
```

Comment:
Worth checking with @tudor-da about the performance implications of doing this.  Recent investigations show doing this sequentially resulted in a 30% slowdown.

Reply:
Yeah, this _can_ reduce throughput by 33% (theoretically) in some cases. But not in the cases that matter imho, where throughput is heavily influenced by the proto marshalling latency (flat/tree transactions and ACS) - those are preoptimized in https://github.com/digital-asset/daml/commit/ce464c242132af889be09e3b4fac255c3a18e772

Reply:
But to be on the safe side, we should run a perf test with this change